third_party/move/move-model/src/metadata.rs

```diff
@@ -0,0 +1,200 @@
+        write!(
+            f,
+            "{}{}",
+            match self {
+                CompilerVersion::V1 => "1",
+                CompilerVersion::V2_0 => "2.0",
+            },
+            if self.unstable() { UNSTABLE_MARKER } else { "" }
+        )
+    }
+}
+
+impl CompilerVersion {
+    /// Return true if this is a stable compiler version. A non-stable version
+    /// should not be allowed on production networks.
+    pub fn unstable(self) -> bool {
+        match self {
+            CompilerVersion::V1 => false,
+            CompilerVersion::V2_0 => true,
+        }
+    }
+
+    /// Check whether the compiler version supports the given language version,
+    /// generates an error if not.
+    pub fn check_language_support(self, version: LanguageVersion) -> anyhow::Result<()> {
+        match self {
+            CompilerVersion::V1 => {
+                if version != LanguageVersion::V1 {
+                    bail!("compiler v1 does only support Move language version 1")
+                } else {
+                    Ok(())
+                }
+            },
+            CompilerVersion::V2_0 => Ok(()),
+        }
+    }
+}
+
+/// Represents a language version.
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, Eq, PartialEq, PartialOrd, Ord)]
+pub enum LanguageVersion {
+    /// The version of Move at around the genesis of the Aptos network end
+    /// of '22. This is the original Diem Move plus the extension of inline
+    /// functions with lambda parameters.
+    V1,
+    /// The upcoming (currently unstable) 2.0 version of Move. The following
+    /// experimental language features are supported so far:
+    ///
+    /// - Access control specifiers as described in AIP-56.
+    /// - receiver style (method style) function calls with auto-referencing
+    V2_0,
+}
+
+impl Default for LanguageVersion {
+    fn default() -> Self {
+        static MOVE_LANGUAGE_V2: Lazy<bool> = Lazy::new(|| read_bool_env_var("MOVE_LANGUAGE_V2"));
+        if *MOVE_LANGUAGE_V2 {
+            Self::V2_0
+        } else {
+            Self::V1
+        }
+    }
+}
+
+impl FromStr for LanguageVersion {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // Strip unstable marker as it is not relevant for parsing.
+        let s1 = s.replace(UNSTABLE_MARKER, "");
+        match s1.as_str() {
+            "1" => Ok(Self::V1),
+            "2" | "2.0" => Ok(Self::V2_0),
+            _ => bail!(
+                "unrecognized language version `{}` (supported versions: `1`, `2`, `2.0`)",
+                s
+            ),
+        }
+    }
+}
+
+impl LanguageVersion {
+    /// Whether the language version is unstable. An unstable version
+    /// should not be allowed on production networks.
+    pub fn unstable(self) -> bool {
+        match self {
+            LanguageVersion::V1 => false,
+            LanguageVersion::V2_0 => true,
+        }
+    }
+}
+
+impl Display for LanguageVersion {
```

Comment:
nit: can we just have some static tables with both display string and boolean for stable/unstable? And avoid all this (and above) logic?

Reply:
We can do this once we have more versions to deal with. It would be an easy local change.